### ISSUE TITLE
ValidateOnChangeOrXXXX is not firing on first event trigger until value change

### ISSUE THREAD

**Issue by bmcdaniels** [aurelia/validation#548]
**I'm submitting a bug report**

* **Library Version:**
2.0.0-rc1

* **JSPM OR Webpack AND Version**
webpack 4.42.1

* **Browser:**
all

* **Language:**
TypeScript 3.8.2

**Current behavior:**
Blur/Focusout does not trigger initial validation until there has been at least one change to the bound value. For instance, 

`<input type="text" value.bind="myVal & validateOnChangeOrBlur" />`

with

```
const rules = ValidationRules
            .ensure((vm: Model) => vm.myVal).required()
```

Step 1: Focus input
Step 2: Blur input (result.. no validation)
Step 3: Focus input
Step 4: Change value
Step 5: Delete value
Step 6: Blur input (result.. validates)

**What is the expected behavior?**
Blur event should fire validation even without a value change otherwise simple validations like `required()` do not function appropriately.

**Comment by Sayan751** [comment]
This is a breaking change introduced in 2.x.
For details refer the release notes: https://github.com/aurelia/validation/releases/tag/2.0.0-rc1

**Comment by Sayan751** [comment]
Refer the original issue for more context: https://github.com/aurelia/validation/issues/509

**Comment by bmcdaniels** [comment]
Perhaps I'm misunderstanding something somewhere. I know that the validation is only supposed to trigger on change only once the blur/focusout occurs, but what I'm seeing is that even the blur/focusout event is being disregarded if there was no value change (I think).

To reproduce, simply set an input to `.required()` and run the application. Focus the input, then blur the input. There will be no validation. Validation only occurs once you've entered some kind of value into the input, then delete it, then blur. 

If the validation binding behavior is just OnBlur, this works as expected. This seems to be exclusive to OnChangeOrBlur.

**Comment by Sayan751** [comment]
@bmcdaniels Your interpretation is correct. The default behavior of `changeOrEVENT` triggers has been changed in the 2.x release.

The primary motivation behind this change is to provide a better UX, based on the "reward early, punish late" principle.

In simple terms this means that when a user focuses on an empty field, and without making any changes blurs, then the field is not "dirty", and therefore there is no need to perform the validation. Also if the user focuses on en empty field, starts typing some invalid value, the change-triggered validation is deferred with the assumption that user has not yet finished typing; therefor the value might be 'incomplete' rather than 'invalid'. With this we end up yelling less at the user.

On the other hand, if the field is validated once, it will be validated on every change. This helps rewarding early. For example, if the field is validated once and found to be invalid, the change-triggered validation will inform the user about the rectification, as soon as the value is valid again.

As the `changeOrEVENT` triggers (`EVENT` can be `blur` or `focusout`) offer two alternatives to perform validation, it is leveraged to have this kind of validation behavior. The `change`, `blur`, or `focuout` triggers, theoretically offer no such flexibility of not performing the validation, as there is one and only one triggering event.

**Comment by itgoran** [comment]
@Sayan751 how to display error message(s) even when user didn't interact with a page (i.e user navigate to a Login page, and fields User name and Password are empty (but required)
How to display validation messages without user need to click on both fields?

**Comment by Sayan751** [comment]
@itgoran I am assuming for any such view, where validation is needed, there is one evident point where we end up calling `validationController.validate()`. That validates all the bindings in the view (as well as the objects registered to the controller for validation). This also further ensures that everything is in order wrt. validation rules defined.

As all the bindings are validated, when this method is called, you have your error message for the field as well. Often, the overall validation is performed using a "submit"-like button, which might be quite usual for your use-case as well in form of "Login" button. 

However, without going into the UI technique too much, let me emphasize that the bottom line is when `validationController.validate()` is called, you get the validation error for the respective fields (assuming the rules are correctly defined for the field and the field binding is marked with `validate` binding behavior). That's how you get the errors w/o interacting with the fields.

Does this answer your question?

**Comment by itgoran** [comment]
@Sayan751 
Thank you for your time and details description.

I've followed your instructions and endup with this (I've included only relevant part of code).

**UserLogin.ts**

```typescript

export class UserLogin {

   constructor(controllerFactory) {
     this.validationController = controllerFactory.createForCurrentScope();
  }

  activate() {
      ProjectValidationRules.SetLoginDataRules(this.loginData);         
      this.validationController.validate({object: this.loginData});
  }
```
**ProjectValidationRules.ts**

```typescript
export class ProjectValidationRules{
   public static SetLoginDataRules(userData: UserLoginData){
         ValidationRules
          .ensure('userName').required()
          .on(userData);
   }
}
```

**user-login.html**

 Here is example of one field with binding

```html
<template>
<div class="form-row">
      <div class="form-group col-md-4" validation-errors.bind="userNameErrors">
        <label for="userName" class.bind="userNameErrors.length ? 'error' : ''"
          t="user_name_label"></label>
        <input type="text" class.bind="userNameErrors.length ? 'error form-control' : 'form-control'" name='userName'
          value.bind='loginData.userName & validate' >
      </div>     
</div>
</template>
```

If I click in field _userName_ it become red, but without interaction it won't.
Note also that I've tried

```typescript
this.validationController.validate({object: this.loginData});
```

and

```typescript
this.validationController.validate();
```

but with no effect.

**Comment by Sayan751** [comment]
@itgoran Although I am not 100% sure, but I think triggering the validation in activate is causing the issue. The elements are probably not attached to DOM yet. You can try that `this.validationController.validate` in attached.

In case it does not work, can you please share a repo/codesandbox/dumber-app that reproduces this behavior?

Sidenote#1: you need to `await` the `validate`
Sidenote#2: showing the error on start seems a bit odd; usual use-case is to trigger validation on "submission".

**Comment by itgoran** [comment]
@Sayan751 
I've postponed validation until submit.
Thank you for effort and time. It was helpful :)

**Comment by Sayan751** [comment]
@itgoran My pleasure :) Glad that it was worked out for you!
